var/spack/repos/builtin/packages/openmpi/package.py

```diff
@@ -354,11 +348,21 @@ class Openmpi(AutotoolsPackage):
 
     conflicts('schedulers=slurm ~pmi', when='@1.5.4:',
               msg='+pmi is required for openmpi(>=1.5.5) to work with SLURM.')
```

Comment:
I think this should be:

```suggestion
    conflicts('schedulers=slurm ~pmi', when='@1.5.4:2',
              msg='+pmi is required for openmpi to work with SLURM.')
    conflicts('schedulers=slurm ~pmi ~pmix', when='@3:',
              msg='+pmi or +pmix is required for openmpi to work with SLURM.')
```

Reply:
I agree this suggested change is more accurate.

Reply:
It seems that this change is causing problems for @haampie 's slurm-based build of openmpi:
```
The application appears to have been direct launched using "srun",
but OMPI was not built with SLURM's PMI support and therefore cannot
execute. There are several options for building PMI support under
SLURM, depending upon the SLURM version you are using:

  version 16.05 or later: you can use SLURM's PMIx support. This
  requires that you configure and build SLURM --with-pmix.

  Versions earlier than 16.05: you must use either SLURM's PMI-1 or
  PMI-2 support. SLURM builds PMI-1 by default, or you can manually
  install PMI-2. You must then build Open MPI using --with-pmi pointing
  to the SLURM PMI library location.

Please configure as appropriate and try again.
```

His previous build had `+pmi` because of the original conflict as written. It might be that newer OpenMPI versions still require `+pmi` for older versions of SLURM: @aweits can you provide clarification on your suggestion?

Reply:
Unfortunately, that will not be true starting with OpenMPI v5.0 as _only_ PMIx will be supported - OMPI will no longer support legacy PMI versions. For older versions of OMPI, you can still use PMI-1 and PMI-2 with legacy PMI so long as you explicitly configure OMPI with those options. The explicit configure is required due to the GPL nature of those legacy PMI libraries - it protects OMPI from the GPL requirements (passing it along to whomever distributes the result).

This is something you may need to consider. If Spack is distributing binary rpms that have OMPI linked to Slurm PMI-1 or PMI-2, then you (Spack) are responsible for the associated GPL conditions. You should probably check with an appropriate legal person as to how the ramifications might impact you.

Reply:
@sethrj did that answer the question? On our side we build slurm as an external w/ pmix, and desire to match that on the openmpi side (i.e. --without-pmi --with-pmix=<path to pmix>).

Reply:
Your last update answers it. In your situation the suggestion is correct but in general it's not. The conflicts should probably be more like:
```python
    conflicts('~pmi', when='^slurm+pmi')
    conflicts('~pmix', when='@3: ^slurm+pmix')
```
The problem with this representation is that I think most users have SLURM installed as an "external" package on an existing server, so Spack has no knowledge of the pmi/pmix variants in use by slurm...